hydro_cli/src/core/gcp.rs

```diff
@@ -308,58 +338,83 @@ impl Host for GCPComputeEngineHost {
-                            {
-                                "image": "debian-cloud/debian-11"
-                            }
-                        ]
-                    }
-                ],
-                "network_interface": [
+
+        let mut tags = vec![];
+        let mut external_interfaces = vec![];
+
+        if self.external_ports.is_empty() {
+            external_interfaces.push(json!({
+                "network": format!("${{google_compute_network.{vpc_network}.self_link}}")
+            }));
+        } else {
+            external_interfaces.push(json!({
+                "network": format!("${{google_compute_network.{vpc_network}.self_link}}"),
+                "access_config": [
                     {
-                        "network": format!("${{google_compute_network.{vpc_network}.self_link}}"),
-                        "access_config": [
-                            {
-                                "network_tier": "STANDARD"
-                            }
-                        ]
+                        "network_tier": "STANDARD"
                     }
                 ]
             }));
 
+            // open the external ports that were requested
+            let open_external_ports_rule = format!("{vm_instance}-open-external-ports");
+            firewall_entries.insert(
+                open_external_ports_rule.clone(),
+                json!({
+                    "name": open_external_ports_rule,
+                    "project": project,
+                    "network": format!("${{google_compute_network.{vpc_network}.name}}"),
+                    "target_tags": [open_external_ports_rule],
+                    "source_ranges": ["0.0.0.0/0"],
+                    "allow": [
+                        {
+                            "protocol": "tcp",
+                            "ports": self.external_ports.iter().map(|p| p.to_string()).collect::<Vec<String>>()
+                        }
+                    ]
+                }),
+            );
+
+            tags.push(open_external_ports_rule);
+
+            resource_batch.terraform.output.insert(
+                format!("{vm_instance}-public-ip"),
+                TerraformOutput {
+                    value: format!("${{google_compute_instance.{vm_instance}.network_interface[0].access_config[0].nat_ip}}")
+                }
+            );
+        }
+
+        resource_batch
+            .terraform
+            .resource
+            .entry("google_compute_instance".to_string())
+            .or_default()
+            .insert(
+                vm_instance.clone(),
+                json!({
+                    "name": vm_instance,
+                    "project": project,
+                    "machine_type": self.machine_type,
+                    "zone": self.region,
+                    "tags": tags,
+                    "metadata": {
+                    "ssh-keys": "hydro:${tls_private_key.vm_instance_ssh_key.public_key_openssh}"
```

Comment:
Could we also expose the path to the ssh keys to python clients?

Reply:
On it.